### DISCUSSION TITLE
GSoC 2023: Monitoring

### DISCUSSION THREAD

**Discussion by linzs148** [pgmoneta/pgmoneta#121] [Ideas]
Hi,

My name is Zhengshun Lin. I am lucky to be selected as a contributor for GSOC 2023 and excited to become one of pgmoneta family.

I will work on implementing monitoring features for pgmoneta. This project involves two parts:
1. expand metrics to help users observe running status of pgmoneta and each server at any time.
2. design a Grafana dashboard to visualize the collected Prometheus metrics.

You can find more details in my proposal:
https://docs.google.com/document/d/1r0Gy3cbdAKevfMZ88iW4GZZx8dz7e5JMwoPh21tWkM0/edit?usp=sharing

If you have any suggestions or questions, please contact me at 1483100349ts@gmail.com.

Hope we have a wonderful time here!

# Progress

## metrics

<table>
<thead>
  <tr>
    <th>aspect</th>
    <th>metric</th>
    <th>description</th>
    <th>state</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td rowspan="10">overall</td>
    <td>pgmoneta_state</td>
    <td>The state of pgmoneta</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_total_space</td>
    <td>The total disk space for pgmoneta</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_used_space</td>
    <td>The disk space used for pgmoneta</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_free_space</td>
    <td>The free disk space for pgmoneta</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_compression</td>
    <td>The compression used (0: None, 1: GZip, 2: ZSTD, 3: LZ4, 4: BZIP2)</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_link</td>
    <td>Use links to limit backup size (0: No, 1: Yes)</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_retention_days</td>
    <td>The retention of pgmoneta in days</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_retention_weeks</td>
    <td>The retention of pgmoneta in weeks</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_retention_months</td>
    <td>The retention of pgmoneta in months</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_retention_years</td>
    <td>The retention of pgmoneta in years</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td rowspan="8">server</td>
    <td>pgmoneta_server_valid</td>
    <td>Is the server in a valid state</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_wal_streaming</td>
    <td>The WAL streaming status of a server</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_retention_server</td>
    <td>The retention a server in days</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_total_size</td>
    <td>The total size for a server</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_server_operation_count</td>
    <td>The count of client operations of a server</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_server_failed_operation_count</td>
    <td>The count of failed client operations of a server</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_server_last_operation_time</td>
    <td>The time of the latest client operation of a server</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_server_last_failed_operation_time</td>
    <td>The time of the latest failed client operation of a server</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td rowspan="12">backup</td>
    <td>pgmoneta_backup</td>
    <td>Is the backup valid for a server</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_backup_size</td>
    <td>The size of a backup for a server</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_backup_elapsed_time</td>
    <td>The backup in seconds for a server</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_backup_retain</td>
    <td>Retain a backup for a server</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_backup_oldest</td>
    <td>The oldest backup for a server</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_backup_newest</td>
    <td>The newest backup for a server</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_backup_count</td>
    <td>The number of valid backups for a server</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_backup_total_size</td>
    <td>The total size of the backups for a server</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_restore_size</td>
    <td>The size of a restore for a server</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_backup_compression_ratio</td>
    <td>The ratio of backup size to restore size for each backup</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_restore_size_increment</td>
    <td>The size increment of a restore for a server</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>pgmoneta_backup_version</td>
    <td>The version of postgresql for a backup</td>
    <td>✔️</td>
  </tr>
  <tr>
    <td>wal</td>
    <td>pgmoneta_wal_total_size</td>
    <td>The total size of the WAL for a server</td>
    <td>✔️</td>
  </tr>
</tbody>
</table>

**Welcome to provide more!** 🍭

### Overall
![overall](https://github.com/pgmoneta/pgmoneta/assets/56420840/5e9df97d-db83-4c49-84e1-41df5c158edb)

### Server
![server](https://github.com/pgmoneta/pgmoneta/assets/56420840/035cf012-8006-464a-8c2d-d012e0f20574) 
![Snipaste_2023-08-19_17-35-04](https://github.com/pgmoneta/pgmoneta/assets/56420840/24d28a7a-0b4b-4d8e-9e71-9b9a5400e86b)

### Backup
![backup](https://github.com/pgmoneta/pgmoneta/assets/56420840/8d7d9d10-9748-4ebc-968e-9914f6386a40)

### WAL
![wal](https://github.com/pgmoneta/pgmoneta/assets/56420840/f89d5caf-41d2-4a80-bb9e-67819258c592)

**Comment by jesperpedersen** [comment]
:+1:

**Comment by resyfer** [comment]
Hi! Nice to meet you. I am Saurav Pal, the GSoC '23 contributor for pgexporter. Since the two projects are very similar (both expose Prometheus metrics), I was really interested about the possible feature addition of Grafana dashboard and your views on tackling it.

On a previous attempt, I had tried connecting pgexporter's exposed metrics to a Grafana dashboard as a data source and the trouble I encountered was that pgexporter did not support custom HTTP queries that Grafana uses to query data sources like `/metrics?query=postgresql_up`, etc. I have seen a very similar [server for exposing metrics](https://github.com/pgmoneta/pgmoneta/blob/main/src/libpgmoneta/prometheus.c#L177) in pgmoneta as well.

I was wondering how you plan to solve this problem, as I really wish to see this feature support on both pgmoneta as well as pgexporter, for which I would gladly port your valuable work from pgmoneta to pgexporter!.

**Reply by linzs148** [reply]
Sorry for the late response. In fact, I encountered an "error reading prometheus: Get /api/v1/query" when importing the /metrics API of pgmoneta directly. My solution is installing prometheus on the server and modify prometheus.yml to encapsulate the original API.
```yml
scrape_configs:
  - job_name: 'pgmoneta'
    metrics_path: '/metrics'
    static_configs:
      - targets: ['localhost:5001']
```
In this way, I can import promethues in grafana (the default port of promethues is 9090) and get metrics of pgmoneta.
![](https://linzs-upload.oss-cn-shanghai.aliyuncs.com/img/Snipaste_2023-06-04_16-22-00.png)

**Reply by resyfer** [reply]
Hi! No problem for the late response! Thanks for sharing this, I'll try it out as soon as I can. Also, please let me know if you plan on adding this in your README, so I can check frequently, and once it's done, I'll add this to `pgexporter`'s README as well giving you proper credits!

**Reply by linzs148** [reply]
> Hi! No problem for the late response! Thanks for sharing this, I'll try it out as soon as I can. Also, please let me know if you plan on adding this in your README, so I can check frequently, and once it's done, I'll add this to `pgexporter`'s README as well giving you proper credits!

Of course I am glad to add this to README. I will comment you once I finish it.

**Reply by linzs148** [reply]
> Hi! No problem for the late response! Thanks for sharing this, I'll try it out as soon as I can. Also, please let me know if you plan on adding this in your README, so I can check frequently, and once it's done, I'll add this to `pgexporter`'s README as well giving you proper credits!

@resyfer Hi Saurav, I have add a document about how to use prometheus and grafana for monitoring [here](https://github.com/pgmoneta/pgmoneta/pull/126/commits/6ebbe7ce1e8e868ce929bae4d3ed57ebf13886a3#diff-4b0327a679cd5fade4c63567f54a34eea312ecbc898341a1509247367855de1c). You are welcome to read it and comment if there is any mistake.

**Reply by resyfer** [reply]
Thanks a lot for this! As soon as your PR gets merged, I'll try to add this to pgexporter as well! I'll try this out on my local machine as well from the instructions.

Also I noticed Ubuntu specific instructions here and there. Though I don't think they're much different from Fedora, but if they are, I'll add Fedora instructions to both pgexporter and pgmoneta READMEs to keep it consistent with the rest of Fedora-like instructions for the rest of the documentation. 🙂

**Comment by solokyo** [comment]
Hi, Zhengshun

I am Jichen, I am one of the GSoC 2023 contributors and I will be working on pgagroal this summer.

I've read your proposal and I have a question.
Will there be any security related metrics? Like number of auth and number of failed auth.

**Reply by linzs148** [reply]
Sorry for the late response. I focused mainly on backup and recovery for the pgmoneta metrics initially, as I think these are core metrics of pgmoneta. Metrics about auth are important as well and I will consider them in my development plan. I will comment you when I updating my peroposal.

**Comment by linzs148** [comment]
@jesperpedersen @MariamFahmy98 I have designed a dashboard for current metrics and [here](https://github.com/pgmoneta/pgmoneta/pull/126) is my pull request. I would like to know your advice for this dashboard design so that I can make it better.

What's more,  in my proposal I divide metrics by attributes (size, count, time and so on), but I noticed that in [pgagroal](https://github.com/agroal/pgagroal/discussions/151) metrics are divided by different parts (pool, server, client and so on). Similarly, I can divide metrics in pggmoneta by different parts (backup, restore, wal so on). I am not sure which way is better and this determines how I will do next. I want to know your ideas.

**Reply by jesperpedersen** [reply]
I would say - look at what information a database administrator would need of information and then split design based on that. So, for the overview page it would be information about number of backups, are they valid, how much disk space have been used, how much disk space is remaining, ... And, then design a page for backup with more detailed information about each backup and the trend (how much are the backup growing, ...) and so on

**Comment by jesperpedersen** [comment]
What are the next steps here ?

**Reply by linzs148** [reply]
I will add more metrics for `backup` and design dashboard for extra metrics.

**Reply by jesperpedersen** [reply]
Sounds good !

Please, post screenshots with descriptions as soon as you have them so we can discuss in more detail. We have about a month left to these enhancements done.

Thanks !
